flink-runtime/src/test/java/org/apache/flink/runtime/state/SharedStateRegistryTest.java

```diff
@@ -143,4 +222,87 @@ public boolean isDiscarded() {
             return discarded;
         }
     }
+
+    private static class KeyedStateHandleWrapper implements KeyedStateHandle {
+        private static final long serialVersionUID = 1L;
+        protected final KeyedStateHandle keyedStateHandle;
+        private volatile boolean isDiscarded;
+
+        KeyedStateHandleWrapper(KeyedStateHandle keyedStateHandle) {
+            this.keyedStateHandle = keyedStateHandle;
+            this.isDiscarded = false;
+        }
+
+        @Override
+        public void registerSharedStates(SharedStateRegistry stateRegistry, long checkpointID) {
+            keyedStateHandle.registerSharedStates(stateRegistry, checkpointID);
+        }
+
+        @Override
+        public KeyGroupRange getKeyGroupRange() {
+            return keyedStateHandle.getKeyGroupRange();
+        }
+
+        @Nullable
+        @Override
+        public KeyedStateHandle getIntersection(KeyGroupRange keyGroupRange) {
+            return keyedStateHandle.getIntersection(keyGroupRange);
+        }
+
+        @Override
+        public void discardState() throws Exception {
+            isDiscarded = true;
+            keyedStateHandle.discardState();
+        }
+
+        @Override
+        public long getStateSize() {
+            return keyedStateHandle.getStateSize();
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            KeyedStateHandleWrapper that = (KeyedStateHandleWrapper) o;
+            return Objects.equals(keyedStateHandle, that.keyedStateHandle);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(keyedStateHandle);
+        }
+
+        boolean isDiscarded() {
+            return isDiscarded;
+        }
+    }
+
+    private static class IncrementalStateHandleWrapper extends KeyedStateHandleWrapper {
```

Comment:
Can you explain why do we need two versions of `KeyedStateHandleWrapper`?
Is it because of `copy`? If so, can't we instead expose wrapped handle via getter and call `copy` on it directly?

Reply:
Actually, the main purpose is for `isDiscarded`.
Current implementation has been changed.

Reply:
But `isDiscarded` is defined in `KeyedStateHandleWrapper`. I'm asking why do we need two classes extending it (`ChangelogStateHandleWrapper` and `IncrementalStateHandleWrapper`).

Reply:
The latest implementation has been changed, there is no `KeyedStateHandleWrapper` anymore.

`IncrementalStateHandleWrapper` is introduced for `#isDiscarded()` and `#copy()`.
`ChangelogStateHandleWrapper` is introduced  for`#isDiscarded()` and `#registerSharedStates` (current `InMemoryChangelogStateHandle` does not register to the state registry).